db/db_versioned.go

```diff
@@ -80,50 +95,229 @@ func (b *BoltDBVersioned) Stop(ctx context.Context) error {
+	buf.Put(ns, versionedKey(key, version), value, fmt.Sprintf("failed to put key %x", key))
+	return b.db.WriteBatch(buf)
 }
 
 // Get retrieves the most recent version
-func (b *BoltDBVersioned) Get(ns string, version uint64, key []byte) ([]byte, error) {
+func (b *BoltDBVersioned) Get(version uint64, ns string, key []byte) ([]byte, error) {
 	if !b.db.IsReady() {
 		return nil, ErrDBNotStarted
 	}
-	// TODO: implement Get
-	return nil, nil
+	// check key's metadata
+	km, err := b.checkNamespaceAndKey(ns, key)
+	if err != nil {
+		return nil, err
+	}
+	hitLast, err := km.checkRead(version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "key = %x", key)
+	}
+	if hitLast {
+		return km.lastWrite, nil
+	}
+	last, v, err := b.get(version, ns, key)
+	if err != nil {
+		return nil, err
+	}
+	hitLast, err = km.deleteInBetween(last, version)
+	if err != nil {
+		return nil, err
+	}
+	if !hitLast && len(v) == 0 {
+		// this is a delete-after-write
+		return nil, ErrDeleted
+	}
+	return v, nil
+}
+
+func (b *BoltDBVersioned) get(version uint64, ns string, key []byte) (uint64, []byte, error) {
+	// construct the actual key = key + version (represented in 8-bytes)
+	// and read from DB
+	var (
+		last  uint64
+		meta  = append(key, 0)
+		value []byte
+	)
+	key = versionedKey(key, version)
+	err := b.db.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(ns))
+		if bucket == nil {
+			return errors.Wrapf(ErrBucketNotExist, "bucket = %x doesn't exist", []byte(ns))
+		}
+		c := bucket.Cursor()
+		k, v := c.Seek(key)
+		if k == nil || bytes.Compare(k, key) == 1 {
+			k, v = c.Prev()
+			if k == nil || bytes.Compare(k, meta) <= 0 {
+				// cursor is at the beginning/end of the bucket or smaller than key's meta
+				panic(fmt.Sprintf("BoltDBVersioned.get(), invalid key = %x", key))
+			}
+		}
+		last = byteutil.BytesToUint64BigEndian(k[len(k)-8:])
+		value = make([]byte, len(v))
+		copy(value, v)
+		return nil
+	})
+	if err == nil {
+		return last, value, nil
+	}
+	if cause := errors.Cause(err); cause == ErrNotExist || cause == ErrBucketNotExist {
+		return 0, nil, err
+	}
+	return 0, nil, errors.Wrap(ErrIO, err.Error())
 }
 
-// Delete deletes a record,if key is nil,this will delete the whole bucket
-func (b *BoltDBVersioned) Delete(ns string, key []byte) error {
+// Delete deletes a record, if key does not exist, it returns nil
+func (b *BoltDBVersioned) Delete(version uint64, ns string, key []byte) error {
 	if !b.db.IsReady() {
 		return ErrDBNotStarted
 	}
-	// TODO: implement Delete
-	return nil
+	// check key's metadata
+	km, err := b.checkNamespaceAndKey(ns, key)
+	if err != nil {
+		return err
+	}
+	if km == nil {
+		return errors.Wrapf(ErrNotExist, "key = %x doesn't exist", key)
+	}
+	if err = km.updateDelete(version); err != nil {
+		return err
+	}
+	if version == km.lastVersion {
+		// write <key, nil> to indicate this is a delete-after-write
```

Comment:
what if the two updates on same version:
```
del k
put k nil
```

Reply:
yes, but this will be a very rare corner case
1. usually a key's value is some valid data
2. if the user intends to use `nil` as a special case, then he can use a special []byte value to avoid the corner case